### PR TITLE
Support the clone system call.

### LINKED ISSUE #155: LKL does not support clone
To fix the layering, we need to return to musl creating threads via the `clone` system call.  Currently, [LKL does not implement `clone` at all](https://github.com/lsds/lkl/blob/21cd7e192cf2a97ea69230b9c0c34da15d157579/arch/lkl/include/asm/syscalls.h#L21).

We need to provide an implementation that handles a the flags required for `pthread_create`.  The correct change in LKL may simply be to provide a host_ops hook that handles clone entirely in the LKL consumer.  The musl implementation of `pthread_create` depends on the following `clone` flags:

 - `CLONE_VM`:  Share address space with the parent. In a single address space world, we cannot support anything other than this.
 - `CLONE_FS`: Share a filesystem namespace with the parent. In a single-process world, this is the obvious thing to do.
 - `CLONE_FILES`: Share a file descriptor table with the parent.  We probably want to support not having this so that our init process can have a separate FD table.
 - `CLONE_SIGHAND`. Share signal handlers. We probably want to support not having this so that our init process can have separate signal handlers.
 - `CLONE_THREAD`. New thread shares PID and has a separate TGID from the parent.  It would be nice to support both variations of this so that we can have a distinct PID for init.
 - `CLONE_SYSVSEM`. Shares ownership of SysV semaphores with the parent.  It doesn't matter too much if we support this because our init process shouldn't use SysV IPC.
 - `CLONE_SETTLS`. Sets the TLS pointer.  Should simply set the %fs base value.
 - `CLONE_PARENT_SETTID`.  Sets the child thread ID in the parent.  Should be easy to support.
 - `CLONE_CHILD_CLEARTID`. Uses the return thread ID as a futex.  See #154.
 - `CLONE_DETACHED`. Has no effect in modern Linux, safe to ignore.

### PR DESCRIPTION
**Pull request by davidchisnall**
This implements two new LKL hooks. The first one to create an lthread
with a specific initial register state (to capture the returns-twice
behaviour of clone, along with the caller's ability to define the stack
and TLS addresses).  The new thread is immediately associated with the
Linux task structure (normally, lthreads are associated with Linux tasks
lazily when they perform a system call).

The second hook destroys a thread.  This is done in response to an exit
system call.  This is somewhat complicated, because LKL never returns to
this thread and the thread's stack may be deallocated by the time we
exit it.

The lthread scheduler does not have an easy way of adding a mechanism to
kill a thread without that thread running.  We can add one eventually,
but for now create a temporary stack that lthreads can use during
teardown and make them run the teardown from there.

Fixes #155

### CODE REVIEW

**Comment by davidchisnall** [comment]
Don't merge this yet: the LKL commit will change once lsds/lkl#1 is merged.  I'll force-push to this branch to update the LKL submodule once that's done.

**Comment by davidchisnall** [comment]
Note that this doesn't yet fix #154.

File: src/include/enclave/lthread.h
Comment (prp) [inline src/include/enclave/lthread.h:210]: Comment?

File: src/include/enclave/lthread.h
Comment (prp) [inline src/include/enclave/lthread.h:267]: Comment?

File: src/include/enclave/lthread.h
Comment (prp) [inline src/include/enclave/lthread.h:279]: Comment?

File: src/lkl/posix-host.c
Comment (prp) [inline src/lkl/posix-host.c:329]: It would be nice to document the parameters, either here or in the header file.

File: src/lkl/posix-host.c
Comment (prp) [inline src/lkl/posix-host.c]:
```suggestion
```

File: src/lkl/posix-host.c
Comment (prp) [inline src/lkl/posix-host.c:372]: A comment would be useful here. Will we ever have to increase this? Perhaps define this as a constant?

File: src/lkl/posix-host.c
Comment (prp) [inline src/lkl/posix-host.c:390]:
```suggestion

```

File: src/sched/lthread.c
Comment (prp) [inline src/sched/lthread.c:631]: As above, it would be good to document these parameters somewhere.

File: tests/basic/clone/clone.s
Comment (prp) [inline tests/basic/clone/clone.s]: Out of curiosity, why do you need assembly to invoke the clone syscall instead of using lkl_syscall directly?

**Review by prp — APPROVED** [review]
LGTM, some minor comments. I see CI failures though...

File: src/lkl/posix-host.c
Comment (davidchisnall) [reply src/lkl/posix-host.c:372]: This is a constant, but I agree on the comment.

File: tests/basic/clone/clone.s
Comment (davidchisnall) [reply tests/basic/clone/clone.s]: It needs to set up the stack explicitly.  The system call returns twice to the same instruction but with the new stack pointer.  You need to then pop the real call address from the stack somehow, which is easy in assembly and impossible to do in C without the compiler being allowed to break it.  Musl and glibc use assembly to call the clone system call directly (rather than via the `syscall` wrapper) for the same reason.

**Comment by davidchisnall** [comment]
CI is failing because the weak symbol for `lkl_syscall` isn't found.  It does for me, so I'm not entirely sure what's going on here...

File: tests/basic/clone/Dockerfile
Comment (SeanTAllen) [inline tests/basic/clone/Dockerfile:1]: out of curiosity, given that 3.6 is rather dated. is there a "no later than X" dependency for sgx-lkl that I'm not aware of?

File: tests/basic/clone/clone.c
Comment (SeanTAllen) [inline tests/basic/clone/clone.c]: why 0 here?

File: tests/basic/clone/Dockerfile
Comment (prp) [reply tests/basic/clone/Dockerfile:1]: We should use a consistent Alpine version in all tests. @letmaik?

File: tests/basic/clone/clone.s
Comment (davidchisnall) [inline tests/basic/clone/clone.s]: Explain why this is needed.

File: tests/basic/clone/Dockerfile
Comment (letmaik) [reply tests/basic/clone/Dockerfile:1]: I opened https://github.com/lsds/sgx-lkl/issues/260. For this PR, it's fine to leave it at an arbitrary version. The [maximum supported](https://github.com/lsds/sgx-lkl/issues/200) currently is 3.10.

File: tests/basic/clone/clone.c
Comment (davidchisnall) [reply tests/basic/clone/clone.c]: Dead code from when I was getting segfaults in the kernel because the stack was misaligned in the clone assembly wrapper.

File: tests/basic/clone/clone.s
Comment (davidchisnall) [reply tests/basic/clone/clone.s]: This code is now in musl, with the relevant comment.

**Comment by SeanTAllen** [comment]
I started looking into the failing `access02` ltp test. Here's how to replicate.

- Start from fresh repo of the `clone` branch.
- make DEBUG=true
- cd tests/ltp/ltp-batch1
- edit ../batch.mk
- comment out the commands for run-hw and run-sw commands (those commands run all tests)

```
run-hw: $(ROOT_FS)
        #@${LTP_TEST_SCRIPT} run-hw

run-sw: $(ROOT_FS)
        #@${LTP_TEST_SCRIPT} run-sw
```

- make clean
- make DEBUG=true
- make run

At this point, you are prepped to run the single test for HW mode:

- SGXLKL_VERBOSE=1 SGXLKL_KERNEL_VERBOSE=0   ../../../build/sgx-lkl-run-oe   --hw-debug   sgxlkl-miniroot-fs.img   /ltp/testcases/kernel/syscalls/access/access02

for SW mode:

SGXLKL_VERBOSE=1 SGXLKL_KERNEL_VERBOSE=0   ../../../build/sgx-lkl-run-oe   --sw-debug   sgxlkl-miniroot-fs.img   /ltp/testcases/kernel/syscalls/access/access02

Result I get
--------------

HW:

tst_test.c:1106: INFO: Timeout per run is 0h 05m 00s
tst_test.c:1125: INFO: No fork support
access02.c:144: PASS: access(file_f, F_OK) as root behaviour is correct.
access02.c:144: PASS: access(file_f, F_OK) as nobody behaviour is correct.
access02.c:144: PASS: access(file_r, R_OK) as root behaviour is correct.
access02.c:144: PASS: access(file_r, R_OK) as nobody behaviour is correct.
access02.c:144: PASS: access(file_w, W_OK) as root behaviour is correct.
access02.c:144: PASS: access(file_w, W_OK) as nobody behaviour is correct.
bad count while changing owner
[[  SGX-LKL ]] FAIL: Kernel panic! Run DEBUG build with SGXLKL_KERNEL_VERBOSE=1 for more information. Aborting...
2020-05-19T19:52:36.000000Z [(H)ERROR] tid(0x7fe8a8ff1700) | :OE_ENCLAVE_ABORTING [/home/sean/openenclave-sgxlkl.git/host/calls.c:oe_call_enclave_function_by_table_id:91]
[   SGX-LKL  ] ethread (4: 19) [   SGX-LKL  ] FAIL: sgxlkl_ethread_init() failed (id=4 result=19 (OE_ENCLAVE_ABORTING))

SW:

It hangs after printing access02.c:144: PASS: access(file_w, W_OK) as nobody behaviour is correct.

----

This is somewhat different than what I am seeing in CI, so, before I proceed any further, @davidchisnall can you try recreating the above?

----

The source of the failing test is https://github.com/linux-test-project/ltp/blob/master/testcases/kernel/syscalls/access/access02.c

**Comment by davidchisnall** [comment]
Thanks @SeanTAllen.  I actually see something more interesting when I try to reproduce this locally:

```
tst_test.c:1106: INFO: Timeout per run is 0h 05m 00s
tst_test.c:1125: INFO: No fork support
access02.c:144: PASS: access(file_f, F_OK) as root behaviour is correct.
access02.c:144: PASS: access(file_f, F_OK) as nobody behaviour is correct.
access02.c:144: PASS: access(file_r, R_OK) as root behaviour is correct.
access02.c:144: PASS: access(file_r, R_OK) as nobody behaviour is correct.
access02.c:144: PASS: access(file_w, W_OK) as root behaviour is correct.
access02.c:144: PASS: access(file_w, W_OK) as nobody behaviour is correct.
Created new host task 7f2fc02f4540 (for 7f2fbeada4c0)
```


So it appears that we're hitting the code path for creating a new clone'd task (for a test that shouldn't be calling clone).  I'll take a look.

**Pull request merged**
